### ISSUE TITLE
Side-by-side comparison FHIR ConceptMap and SSSOM

### ISSUE THREAD

**Issue by matentzn** [mapping-commons/sssom#170]
https://www.hl7.org/fhir/conceptmap.html and SSSOM are obviously very different in scope, but it would be useful to explore commonalities and places for alignment. @DaveraGabriel is working with @joeflack4 on a cursory alignment between the two, and it would be great if you could continue to update us on the progress.

Some notes from watching Grahame Grieves [presentation](https://www.youtube.com/watch?v=KFYQf1MZPD4) on the FHIR mapping language:

- In general, ConceptMap is not designed to be a framework for sharing _simple_ terminological mappings. It is designed to provide not only structural mappings between information entities (classes, instances, etc), but also define the transformation process between two conceptual schemas. Indeed, from listening to Grahame, this seems to be be main purpose of conceptmap: providing a language for defining the transformation for one model to another, including transforms at concept level, and then sharing these transforms in a way that a resource provide can simply run the transform locally (i assume the FHIR servers would also allow to automatically transform, but I have not looked at that yet). So, FHIR ConceptMap is _much_ more comprehensive, and serves a very different use case then SSSOM. That said, some commonalities on the concept level can be compared.  
- Mapping "equivalence" (relatedto, equivalent, equal, wider, narrower, subsumes, specialises) should be mapped carefully to semantic properties in skos and owl.
- There is a bit of "display" related information in the data model, like `div` element.
- TBD

**Comment by matentzn** [comment]
In this video https://www.youtube.com/watch?v=2_rg79JKsP8 it becomes even more clear:

- mapping files are specified in a way that they can be read by a mapping engine _to transform one datasource to another_.

**Comment by joeflack4** [comment]
Thanks Nico. I read through your comments, and I added links to https://github.com/HOT-Ecosystem/hapi-fhir-jpaserver-starter/issues/3.

Perhaps we should use Wiki or static `.md` files to store these notes if they flesh out more.

**Comment by joeflack4** [comment]
@matentzn RE:
> conceptmap: providing a language for defining the transformation for one model to another, including transforms at concept level, and then sharing these transforms in a way that a resource provide can simply run the transform locally

You may be confused with StructureMap. From the [ConceptMap page](https://www.hl7.org/fhir/conceptmap.html), where the word "transform" only appears twice, in the concept of StructureMap:
> In addition to ConceptMap, there is also the [StructureMap](https://www.hl7.org/fhir/structuremap.html) resource. The ConceptMap resource defines relationships between concepts in their own right, along with grading of their equivalencies, while the StructureMap defines an executable transform for instances that conform to a known structure.
> Note that this is different from [StructureMap](https://www.hl7.org/fhir/structuremap.html) because that is intended to define an executional transform between structures, not a conceptual model.

As for ConceptMap, the short definition on that page is as follows:
> A statement of relationships from one set of concepts to one or more other concepts - either concepts in code systems, or data element/data element concepts, or classes in class models.

**Comment by matentzn** [comment]
Ok, maybe you are right. I was going by the presentation I was watching, and maybe I somehow didn't catch this structuremap vs conceptmap distinction and its relationship to the FHIR mapping language. So what exactly is conceptmap that: an API spec? A datamodel for mappings at concept level? If so, maybe someone that is a bit more into the FHIR stack could try a brief comparison of SSSOM vs ConceptMap as they see it.

**Comment by joeflack4** [comment]
Davera and I are in an "ask an expert" session right now at the HL7 Connectathon. Based on what Rob McClure said, I think we can continue with our plans to extend ConceptMap.

I wanted to ask him to tell me generally about the FHIR mapping language, though, but he doesn't really know about it.

I feel like it's not a stretch to call ConceptMap (and the other FHIR resources) a data model. But really, FHIR's main concern is more about the transfer (API) layer.

Rob did give the history of ConceptMap, and it was designed particularly for the use case of mapping value sets, which involves mapping concepts between those value sets. And as such, a value set URI is an optional parameter that you can include, which helps provide context to the values in the value set. But it is still a valid use case to use ConceptMap at the concept to concept level, outside of the context of their usages in value sets.

I've brought up SSSOM in probably 3-4 "ask an expert" sessions so far, with people like Rob McClure, Graham Grieve, Peter Jordan, and Reuben Daniels. I never asked any of them to specifically compare both, as none of them were prior familiar with SSSOM, and it would take a lot of time in these sessions to give a thorough introduction to SSSOM. However, I did familiarize them informally by sharing my screen. I showed them the SSSOM and ConceptMap specs side-by-side, and gave them examples of how we're thinking about doing the extension. They seemed to understand and did not object; so it seems to me implicit that SSSOM makes sense as a use case for extending ConceptMap.

We could always (a) make an API just for SSSOM, (b) or make an SSSOM resource, but I think (c) extending ConceptMap is also fine. I don't have a strong preference, but I'm fine with continuing our plan to go with (c). Maybe @DaveraGabriel has a rationale for choosing (c) over (a) or (b).

**Comment by joeflack4** [comment]
I don't have any qualms with anything you mentioned other than potentially the usage of the metadata resource. I just became familiar with it the other day, but it seems analogous to the servers capability statement. So it seems to be more a large description of metadata about what the server is capable of, rather than anything to do with concept maps. I'm wondering why he suggested this. Or maybe he didn't mean to say the metadata resource, but meant to specifically speak about how to add additional metadata to an existing resource.

**Comment by DaveraGabriel** [comment]
@joeflack4 fair enough.  I haven't looked at the metadata resource at all, but was taking Michael's comment on face value.  NO matter where we land, I think it would be valuable to implement an SSSOM extension to work with / test out

**Comment by cmungall** [comment]
is this the best URL for ConceptMap2? http://hl7.org/fhir/5.0.0-snapshot1/conceptmap2.html

**Comment by DaveraGabriel** [comment]
Hello @cmungall 
Correct... with some caveats
There are some options for the SSSOM on FHIR development based on whether FHIR R4 or R5 ConceptMap resource is selected as a base to develop an extension.  The link you provide above is for ConceptMap2 a component of FHUR R5, and I do think that's the best place to start becuase there are significant differences between R4 and R5 concept map resources and its where things are moving... eventually.  The FHIR management group had a hard think about its timeline to move to R5 for FHIR just prior to the May Ballot [as outlined here](https://onfhir.hl7.org/2022/02/28/soliciting-feedback-concerning-the-roadmap-for-fhir-r5/) and the decision was to go with a hybrid of choice #3 outlined [in this document](https://docs.google.com/document/d/1hhPwyuXxsiwQMVBwu3fCDsPgQgHVkhr1mUm8KcVUUaQ/edit). 

When I created the suggested SSSOM extension, I used the ConceptMap2 resource available at the link you cite above.  However, there may be the necessity to also do an R4 ConceptMap resource extension given we will be in this liminal space with the FHIR R4 / R5 version status for some time to come.  Since we need to get started with our work, I think its a good bet to start with where the "ball will be going" but its good to also be aware of these other developments in the interim.  There is not one solid answer to your question... but go with r5 / ConceptMap2 for now

**Comment by joeflack4** [comment]
Just adding a link to the TIMS SSSOM Extension GH issue: https://github.com/HOT-Ecosystem/hapi-fhir-jpaserver-starter/issues/3
And related `sssom-py` PR: https://github.com/mapping-commons/sssom-py/pull/250

**Comment by matentzn** [comment]
Moved this to https://github.com/mapping-commons/rda-fair-mappings/issues/49 - out of scope for SSSOM project
